---
title: "Class 15"
author: "Yuqi Zhang(PID 15655958)"
date: "3/8/2022"
output: pdf_document
---

##1. Investigating pertussis cases by year

>Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
cdc<- data.frame(
                          Year = c(1922L,
                                   1923L,1924L,1925L,1926L,1927L,1928L,
                                   1929L,1930L,1931L,1932L,1933L,1934L,1935L,
                                   1936L,1937L,1938L,1939L,1940L,1941L,
                                   1942L,1943L,1944L,1945L,1946L,1947L,1948L,
                                   1949L,1950L,1951L,1952L,1953L,1954L,
                                   1955L,1956L,1957L,1958L,1959L,1960L,
                                   1961L,1962L,1963L,1964L,1965L,1966L,1967L,
                                   1968L,1969L,1970L,1971L,1972L,1973L,
                                   1974L,1975L,1976L,1977L,1978L,1979L,1980L,
                                   1981L,1982L,1983L,1984L,1985L,1986L,
                                   1987L,1988L,1989L,1990L,1991L,1992L,1993L,
                                   1994L,1995L,1996L,1997L,1998L,1999L,
                                   2000L,2001L,2002L,2003L,2004L,2005L,
                                   2006L,2007L,2008L,2009L,2010L,2011L,2012L,
                                   2013L,2014L,2015L,2016L,2017L,2018L,
                                   2019L),
  No..Reported.Pertussis.Cases = c(107473,
                                   164191,165418,152003,202210,181411,
                                   161799,197371,166914,172559,215343,179135,
                                   265269,180518,147237,214652,227319,103188,
                                   183866,222202,191383,191890,109873,
                                   133792,109860,156517,74715,69479,120718,
                                   68687,45030,37129,60886,62786,31732,28295,
                                   32148,40005,14809,11468,17749,17135,
                                   13005,6799,7717,9718,4810,3285,4249,
                                   3036,3287,1759,2402,1738,1010,2177,2063,
                                   1623,1730,1248,1895,2463,2276,3589,
                                   4195,2823,3450,4157,4570,2719,4083,6586,
                                   4617,5137,7796,6564,7405,7298,7867,
                                   7580,9771,11647,25827,25616,15632,10454,
                                   13278,16858,27550,18719,48277,28639,
                                   32971,20762,17972,18975,15609,18617)
)
```


```{r}
library(ggplot2)
baseplot <- ggplot(cdc) +
  aes(Year,No..Reported.Pertussis.Cases ) +
  geom_point() +
  geom_line() +
  labs(x="Year", y = "Number of classes", title ="Pertussis Cases by Year (1922-2019)" )
  
  baseplot
```

##2. A tale of two vaccines (wP & aP)

Two types of pertussis vaccines are currently available: whole-cell pertussis (wP) and acellular pertussis (aP). The first vaccines were composed of ‘whole cell’ (wP) inactivated bacteria. The latter aP vaccines use purified antigens of the bacteria (the most important pertussis components for our immune system). 

>Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
baseplot +
  geom_vline(xintercept=1946, col = "blue", lty = "dashed")+
  geom_vline(xintercept=1996, col = "red" ,lty = "dashed")
```
Before the introduction of vaccines, many people are infected by pertussis each year. After the appearance of whole-cell Pertussis vaccine, the number of cases decreased significantly and there is nearly no patient from 1975 to 2000. After the introduction of acellular pertussiss vaccine, the number of cases increased a little but it is much lesser than the time without any vaccine.

>Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

It is clear from the CDC data that pertussis cases are once again increasing. The pertussis field has several hypotheses for the resurgence of pertussis including (in no particular order): 1) more sensitive PCR-based testing, 2) vaccination hesitancy 3) bacterial evolution (escape from vaccine immunity), 4) waning of immunity in adolescents originally primed as infants with the newer aP vaccine as compared to the older wP vaccine.

## 3.  Exploring CMI-PB data

```{r}
library(jsonlite)
```

```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```

```{r}
head(subject, 3)
```

> Q4 How may aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

> How many total

```{r}
nrow(subject)
```

> Q5 How many Male and Female subjects/patients are in the dataset?

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$biological_sex,subject$race)
```

```{r}
library(lubridate)
today()
```
How many days have passed since new year 2000

```{r}
today() - ymd("2000-01-01")
```
What is this in years?

```{r}
time_length( today() - ymd("2000-01-01"),  "years")
```


> Q7/8 are optional

>Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```


# Joining Multiple tables

Read the specimen and ab_titer tables into R and store the data as specimen and titer named data frames.


```{r}
# Complete the API URLs...
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 

```

Have a quick look 
```{r}
head(subject,3)
```

```{r}
head(specimen, 3)
```

>Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

I need to use `inner_join()` here

```{r}
dim(subject)
```
```{r}
dim(specimen)
```


```{r}
library(dplyr)
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

>Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.


```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

>Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```
>Q12. What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit)
```

Because 8 is an ongoing project, it has much fewer number of visit compared to others.


##4. Examine IgG1 Ab titer levels

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1,3)
```

>Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
```

>Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?
 
Antigen FIM2/3, FHA, PRN show differences becuases these antibodies are in the vacccine. The acellular vaccine contains Pertactin, Fimbriae, FHA, and Weakened pertussis toxin. 


We can attempt to examine differences between wP and aP here by setting color and/or facet values of the plot to include infancy_vac status (see below). However these plots tend to be rather busy and thus hard to interpret easily.


```{r}
ggplot(ig1) +
  aes(ab_titer, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  theme_bw()
```
>Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).


```{r}
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(ig1, antigen== "FIM2/3") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

>Q16. What do you notice about these two antigens time course and the FIM2/3 data in particular?

The controll antigen(Measles) barely changed while FIM2/3 levels increases over time and exceed those of Measles. FIM2/3 also appear to peak at visit 5 and then decline. This trend appears similar for both wP and aP subjects.

>Q17. Do you see any clear difference in aP vs. wP responses?

For the antigen of FIM2/3, I can see some differences in  aP vs. wP responses in visit 2, 4, 6, and 7, but overall they are pretty similar to me.

##5. Obtaining CMI-PB RNASeq data

Pull RNA-Seq Data from the CMI-PB

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

To facilitate further analysis we need to “join” the rna expression data with our metadata meta, which is itself a join of sample and specimen data. This will allow us to look at this genes TPM expression values over aP/wP status and at different visits (i.e. times):

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

>Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```
>Q19.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

The expression of the gene increase from 1-4 visit, and reaches the peak at 4 visit. Then, the expression of gene decreased back at visit 5.

>Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

This pattern in time does not match the trend of antibody titer data, because cells make antigens which can live longer. Thus, the antibody will increase a little bitter later than the cells.


We can dig deeper and color and/or facet by infancy_vac status:
```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```
There is no obvious wP vs. aP differences here even if we focus in on a particular visit:

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

##Working with long format data

```{r}
# Change for your downloaded file path
rnaseq <- read.csv("~/Downloads/2020LD_rnaseq.csv")

head(rnaseq,3)
dim(rnaseq)
```

```{r}
n_genes <- table(rnaseq$specimen_id)
head( n_genes , 10)
```

How many specimens?

```{r}
length(n_genes)
```

Check if there are the same number of genes for each specimen
```{r}
all(n_genes[1]==n_genes)
```

All looks good at this stage so let’s convert to wider format with the pivot_wider() function from the tidyr package:

```{r}
library(tidyr)

rna_wide <- rnaseq %>% 
  select(versioned_ensembl_gene_id, specimen_id, tpm) %>%
  pivot_wider(names_from = specimen_id, values_from=tpm)

dim(rna_wide)
```

```{r}
head(rna_wide[,1:7], 3)
```

